h2o-algos/src/main/java/hex/tree/isoforextended/ExtendedIsolationForest.java

```diff
@@ -0,0 +1,321 @@
+//
+//                    if ((2 * i + 1) < nodes.length) {
+//                        nodes[2 * i + 1] = new Node(left, currentHeight);
+//                        nodes[2 * i + 2] = new Node(right, currentHeight);
+//                    }
+//                }
+//            }
+//        }
+//
+//        public void print() {
+//            for (int i = 0; i < 16; i++) {
+//                if (nodes[i] == null)
+//                    System.out.print("null ");
+//                else
+//                    System.out.print(nodes[i].size + " ");
+//            }
+//            System.out.println("");
+//        }
+//
+//        public double computePathLength(Vec row) {
+//            int position = 0;
+//            Node node = nodes[0];
+//            double score = 0;
+//            while (!node.external) {
+//                Vec sub = MatrixUtils.subtractionVtv(row, node.p);
+//                double mul = MatrixUtils.productVtV(sub, node.n);
+//                if (mul <= 0) {
+//                    position = 2 * position + 1;
+//                } else {
+//                    position = 2 * position + 2;
+//                }
+//                if (position < nodes.length)
+//                    node = nodes[position];
+//                else
+//                    break;
+//            }
+//            score += node.currentHeight + avgPathLengthUnsucSearch(node.size);
+//            return score;
+//        }
+//    }
+
+
+    private class FilterLteTask extends MRTask<FilterLteTask> {
+
+        private double value;
+        private Vec mul;
+
+        public FilterLteTask(Vec vec, double value) {
+            this.mul = vec;
+            this.value = value;
+        }
+
+        @Override
+        public void map(Chunk[] cs, NewChunk[] ncs) {
+            for (int row = 0; row < cs[0]._len; row++) {
+                double num = mul.at(cs[0].start() + row);
+                if (num <= value) {
+                    for (int column = 0; column < cs.length; column++) {
+                        ncs[column].addNum(cs[column].atd(row));
+                    }
+                }
+            }
+        }
+    }
+
+    private class FilterGtRightTask extends MRTask<FilterGtRightTask> {
+
+        private double value;
+        private Vec mul;
+
+        public FilterGtRightTask(Vec mul, double value) {
+            this.mul = mul;
+            this.value = value;
+        }
+
+        @Override
+        public void map(Chunk[] cs, NewChunk[] ncs) {
+            for (int row = 0; row < cs[0]._len; row++) {
+                double num = mul.at(cs[0].start() + row);
+                if (num > value) {
+                    for (int column = 0; column < cs.length; column++) {
+                        ncs[column].addNum(cs[column].atd(row));
+                    }
+                }
+            }
+        }
+    }
+
+    private class Node {
+        private Frame frame;
```

Comment:
Try to avoid saving the whole `Frame.` Save the key of the frame instead. When you need to use the frame, call something like `DKV.get(key)`.

Reply:
Done